### DISCUSSION TITLE
Multiple requests are normal?

### DISCUSSION THREAD

**Discussion by maiconsanson** [amannn/next-intl#1085] [Q&A]
Hi folks,

Testing my error pages I realized getting 4 requests for each not-found path in production.
I followed these instructions: https://next-intl-docs.vercel.app/docs/environments/error-files

For example, I only accessed `teste` and `asdasd` once:

<img width="1239" alt="Screenshot 2024-05-22 at 18 18 26" src="https://github.com/amannn/next-intl/assets/1965455/04809a70-8318-4a30-ad58-061ace0b7d1f">

I'm seeing multiple requests for the same URL in other situations too.

Somebody already faced something like this? I don't think 4 requests are normal for the same 404 error page.
This project isn't public yet so I'm afraid about the high Vercel pricing when launched.

My middleware:

```jsx
const handleI18nRouting = createIntlMiddleware({
  defaultLocale: 'pt',
  localePrefix: 'always',
  locales: Object.keys(LOCALES),
})

export default async function middleware(request) {
  const response = handleI18nRouting(request)
  const { cookies } = request || {}
  const { pathname } = request.nextUrl || {}

  const redirectTo = (route) => {
    const splitedPathname = pathname.split('/')
    const locale = splitedPathname[1]

    if (!locale) return response

    const url = new URL(`/${locale}/${route}`, request.url).toString()
    return NextResponse.redirect(url)
  }

  const token = cookies.get(COOKIES.TOKEN)?.value
  const user = isCryptoTokenValid(token) ? await getUserData(token) : null
  const { cookie_expires_at, new_token, status } = user || {}

  if (new_token) {
    response.cookies.set(COOKIES.TOKEN, new_token, {
      expires: new Date(cookie_expires_at),
      httpOnly: true,
      secure: true,
    })
  }

  const isLogoutPage = checkRouteType(pathname, 'logout')
  const isAuthPage = checkRouteType(pathname, 'auth') && !isLogoutPage
  const isPrivatePage = checkRouteType(pathname, 'private')

  const isAuthenticated = user ? canLogin({ status }) : false
  if (isAuthPage && isAuthenticated) return redirectTo(routes.stories.featured)
  if (isPrivatePage && !isAuthenticated) return redirectTo(routes.auth.login)

  return response
}

export const config = {
  // Match only internationalized pathnames
  matcher: ['/', '/(en|pt)/:path*'],
}
```

**Comment by amannn** [comment]
Is it expected that you're seeing this error? Please see [the troubleshooting guide in the docs](https://next-intl-docs.vercel.app/docs/routing/middleware#unable-to-find-locale).

> I'm seeing multiple requests for the same URL

What you're seeing above are log entries, I wouldn't say that these are strictly the same as requests. It might be that the renderer attempts to render a few components in parallel (4 in your case, could also include `generateMetadata` etc.) and in each of those there are calls to `next-intl`.

**Reply by maiconsanson** [reply]
I see, thanks for the reply.

This error is seen for a default not found page when the matcher didn't match a request but I'm not using a component with some next-intl API. For example `/asdas` .

I'm using the default `src/app/not-found` with the `src/app/layout` without metadata.

```
"use client";

import Error from "next/error";

export default function NotFound() {
  return (
    <html lang="en">
      <body>
        <Error statusCode={404} />
      </body>
    </html>
  );
}
```


```
export default function RootLayout({ children }) {
  return children;
}
```

Is it normal? I can create a reproduction link if required.

**Reply by amannn** [reply]
Hmm, same as in [`example-app-router`](https://github.com/amannn/next-intl/tree/main/examples/example-app-router)? I can't reproduce this there (also after [commenting this out](https://github.com/amannn/next-intl/blob/d2668b856468b7203dd9e16cf134ea0219e59b39/examples/example-app-router/src/middleware.ts#L20-L22)).

Can you find a difference with your implementation?

**Reply by maiconsanson** [reply]
I think I got it. In your example, you've set the `unstable_setRequestLocale` in both (`[locale]/layout` and `[locale]/page`).
The error occurs if I disable redirects that add missing locales in middleware and don't have the `unstable_setRequestLocale`.

```Unable to find next-intl locale because the middleware didn't run on this request. See https://next-intl-docs.vercel.app/docs/routing/middleware#unable-to-find-locale. The notFound()```.

Is this behavior expected to be?
